### ISSUE TITLE
daemon dies on system bus

### ISSUE THREAD

**Issue by flihp** [01org/tpm2-abrmd#203]
I've just run into a weird issue where the daemon appears to be getting kicked off the dbus system bus when clients execute a command, but it works fine on the session bus. I'm collecting data still and trying to isolate the change that introduced the issue (if there is one).

**Comment by flihp** [comment]
The failure occurs after tabrmd switched from using pipes to sockets. Still the issue only happens when the daemon is on the system bus. Session bus works fine. This makes me think it's not in the code here but something to do with the difference between the two bus types. Not sure how sockets vs pipes plays in to this.

**Comment by flihp** [comment]
Oh SELinux, how I've missed you. Problem goes away with SELinux disabled. No AVCs in in the audit log either. The tabrmd is running in the `unconfined_service_t` domain. The change causing the issue is the switch from using pipes to sockets so `unconfined_service_t` and my client domain must be allowed to use pipes but not sockets.

**Comment by flihp** [comment]
Any suggestions @martinezjavier? I can work around this easy enough locally but would be interested in your thoughts on a more permanent fix. Instructing people to put SELinux into permissive mode isn't an option. Creating a type for the tabrmd is probably the right thing to do here.

**Comment by tstruk** [comment]
sorry, I didn't realize the '#' at the beginning of a line makes it look like this.

**Comment by flihp** [comment]
well you found the github caps lock :smile: 
I'm gonna take a swing at implementing a unique SELinux domain for the daemon. We're gonna need it anyways. It'll have to go upstream to the refpolicy eventually.

**Comment by martinezjavier** [comment]
@tstruk  I don't think your suggestion will work, since the AVC denial is not for the tpm2-abrmd process but for for the dbus-daemon:

```
type=AVC msg=audit(1507198100.003:2076): avc:  denied  { read write } for  pid=1076 comm="dbus-daemon" path="socket:[217760]" dev="sockfs" ino=217760 scontext=system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:unconfined_service_t:s0 tclass=unix_stream_socket permissive=0
```

In fact, since the tpm2-abrmd is in an unconfined domain, the SELinux rules don't apply to it.

My understanding is that for the system bus, the dbus-daemon isn't allowed to read/write sockets created by processes in an unconfined domain. It works on the session bus because dbus-daemon also runs on an unconfined domain in that case:

```
$ ps -efZ | grep dbus-daemon
system_u:system_r:system_dbusd_t:s0-s0:c0.c1023 dbus 1076 1  0 01:33 ? 00:00:15 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
unconfined_u:unconfined_r:unconfined_dbusd_t:s0-s0:c0.c1023 gdm 1349 1316  0 01:33 ? 00:00:00 /usr/bin/dbus-daemon --session --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
```

So the actual Access Vector rule that will make it "work" is:

```
allow system_dbusd_t unconfined_service_t:unix_stream_socket { read write };
```

But that is obviously wrong. I agree with @flihp that the correct approach is to create a domain for the tabrmd so it doesn't run in an unconfined domain.

I'm very far from being an SELinux expert though, so I may be wrong on my assumptions.

**Comment by flihp** [comment]
I took a few hours yesterday to re-learn some SELinux stuff. It's been awhile, the tooling has changed a little bit and I'm not familiar with the Fedora package layout. I've got a unique domain set up for the tabrmd and it can connect to the system dbus with the policy enforced. I'll run through the test cases later today to see what other permissions it needs.

Integrating the SELinux policy build process into our build is another story. For the short term (to unblock use on Fedora) the policy will probably live in a `contrib` directory until we can figure out the build stuff.

**Comment by martinezjavier** [comment]
@flihp don't worry about the policy making it to Fedora (either as part of the tpm2-abrmd package or the system wide selinux-policy package), I can help with that. Also not a SELinux expert as mentioned but at I've a lot of people to ask :smiley:

**Comment by flihp** [comment]
Thanks for that @martinezjavier. I'll try to send this upstream eventually but that requires testing with the upstream refpol and I don't have the cycles for it ATM. I'll create a new ticket to track that work.
